src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs

```diff
@@ -420,8 +417,7 @@ public void SetResult(TResult result)
             // If there isn't one, get a task and store it.
             if (m_task is null)
             {
-                m_task = GetTaskForResult(result);
-                Debug.Assert(m_task != null, $"{nameof(GetTaskForResult)} should never return null");
+                m_task = Threading.Tasks.Task.FromResult(result);
```

Comment:
@jkotas, is this going to hit inlining problems because of one generic type calling into another generic method?

Reply:
I think so.

Reply:
Thanks.  Hmm, I just looked at an example, and it actually looks like it's successful, for both value and reference TResults.  Maybe your recent change helped here?

Reply:
Could you please share your test?

The caller of `SetResult` has to be generic non-inlineable method too for it to hit the inlining limitations.

Reply:
> Could you please share your test?

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System.Threading.Tasks;

[DisassemblyDiagnoser(maxDepth: 5)]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private bool _value;

    [Benchmark]
    public async Task<bool> Test1(bool value) => _value;

    [Benchmark]
    public async Task<string> Test2() => "";

    [Benchmark]
    public async Task<int> Test3() => 0;
}
```
run with:
```
dotnet run -c Release -f net5.0 --corerun d:\coreclrtest\pr\corerun.exe --filter **
```
where `d:\coreclrtest\pr\` is a build with my changes.

Here are the relevant methods disassembled...

For Test1:
```assembly
; Program+<Test1>d__2.MoveNext()
       push      rbp
       push      rsi
       sub       rsp,28
       lea       rbp,[rsp+30]
       mov       [rbp+0FFF0],rsp
       mov       [rbp+10],rcx
       mov       rcx,[rbp+10]
       mov       rcx,[rcx]
       movzx     edx,byte ptr [rcx+8]
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx+8],0FFFFFFFE
       mov       rsi,[rbp+10]
       add       rsi,10
       mov       rcx,[rsi]
       test      rcx,rcx
       jne       short M03_L02
       test      edx,edx
       jne       short M03_L00
       mov       rcx,7FF7FFBB0020
       mov       edx,351
       call      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       mov       rcx,1E3647A19C8
       mov       rdx,[rcx]
       jmp       short M03_L01
M03_L00:
       mov       rcx,7FF7FFBB0020
       mov       edx,351
       call      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       mov       rdx,1E3647A19C0
       mov       rdx,[rdx]
M03_L01:
       mov       rcx,rsi
       call      CORINFO_HELP_CHECKED_ASSIGN_REF
       jmp       short M03_L03
M03_L02:
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Boolean, System.Private.CoreLib]].SetExistingTaskResult(System.Threading.Tasks.Task`1<Boolean>, Boolean)
M03_L03:
       nop
       lea       rsp,[rbp+0FFF8]
       pop       rsi
       pop       rbp
       ret
       push      rbp
       push      rsi
       sub       rsp,28
       mov       rbp,[rcx+20]
       mov       [rsp+20],rbp
       lea       rbp,[rbp+30]
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx+8],0FFFFFFFE
       mov       rcx,[rbp+10]
       add       rcx,10
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Boolean, System.Private.CoreLib]].SetException(System.Exception)
       lea       rax,[7FF7FFE062A0]
       add       rsp,28
       pop       rsi
       pop       rbp
       ret
; Total bytes of code 209
```

For Test2:
```assembly
; Program+<Test2>d__3.MoveNext()
       push      rbp
       push      rdi
       push      rsi
       push      rbx
       sub       rsp,28
       lea       rbp,[rsp+40]
       mov       [rbp+0FFE0],rsp
       mov       [rbp+10],rcx
       mov       rcx,28090003060
       mov       rsi,[rcx]
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx],0FFFFFFFE
       mov       rdi,[rbp+10]
       add       rdi,8
       mov       rdx,[rdi]
       test      rdx,rdx
       jne       short M03_L02
       test      rsi,rsi
       jne       short M03_L00
       mov       rcx,7FF7FFB70020
       mov       edx,2B
       call      CORINFO_HELP_CLASSINIT_SHARED_DYNAMICCLASS
       mov       rcx,28090002E30
       mov       rdx,[rcx]
       jmp       short M03_L01
M03_L00:
       mov       rcx,offset MT_System.Threading.Tasks.Task`1[[System.String, System.Private.CoreLib]]
       call      CORINFO_HELP_NEWSFAST
       mov       rbx,rax
       mov       dword ptr [rbx+34],1000000
       lea       rcx,[rbx+38]
       mov       rdx,rsi
       call      CORINFO_HELP_ASSIGN_REF
       mov       rdx,rbx
M03_L01:
       mov       rcx,rdi
       call      CORINFO_HELP_CHECKED_ASSIGN_REF
       jmp       short M03_L03
M03_L02:
       mov       r8,rsi
       mov       rcx,offset MT_System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.String, System.Private.CoreLib]]
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib]].SetExistingTaskResult(System.Threading.Tasks.Task`1<System.__Canon>, System.__Canon)
M03_L03:
       nop
       lea       rsp,[rbp+0FFE8]
       pop       rbx
       pop       rsi
       pop       rdi
       pop       rbp
       ret
       push      rbp
       push      rdi
       push      rsi
       push      rbx
       sub       rsp,28
       mov       rbp,[rcx+20]
       mov       [rsp+20],rbp
       lea       rbp,[rbp+40]
       mov       r8,rdx
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx],0FFFFFFFE
       mov       rcx,[rbp+10]
       add       rcx,8
       mov       rdx,offset MT_System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.String, System.Private.CoreLib]]
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib]].SetException(System.Exception)
       lea       rax,[7FF7FFDC6728]
       add       rsp,28
       pop       rbx
       pop       rsi
       pop       rdi
       pop       rbp
       ret
; Total bytes of code 251
```

For Test3:
```assembly
; Program+<Test3>d__4.MoveNext()
       push      rbp
       push      rsi
       sub       rsp,28
       lea       rbp,[rsp+30]
       mov       [rbp+0FFF0],rsp
       mov       [rbp+10],rcx
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx],0FFFFFFFE
       mov       rsi,[rbp+10]
       add       rsi,8
       mov       rcx,[rsi]
       test      rcx,rcx
       jne       short M03_L00
       mov       rcx,7FF7FFB70020
       mov       edx,351
       call      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       mov       rdx,1C9100019D0
       mov       rdx,[rdx]
       cmp       dword ptr [rdx+8],1
       jbe       short M03_L02
       mov       rdx,[rdx+18]
       mov       rcx,rsi
       call      CORINFO_HELP_CHECKED_ASSIGN_REF
       jmp       short M03_L01
M03_L00:
       xor       edx,edx
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Int32, System.Private.CoreLib]].SetExistingTaskResult(System.Threading.Tasks.Task`1<Int32>, Int32)
M03_L01:
       nop
       lea       rsp,[rbp+0FFF8]
       pop       rsi
       pop       rbp
       ret
M03_L02:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
       push      rbp
       push      rsi
       sub       rsp,28
       mov       rbp,[rcx+20]
       mov       [rsp+20],rbp
       lea       rbp,[rbp+30]
       mov       rcx,[rbp+10]
       mov       dword ptr [rcx],0FFFFFFFE
       mov       rcx,[rbp+10]
       add       rcx,8
       call      System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Int32, System.Private.CoreLib]].SetException(System.Exception)
       lea       rax,[7FF7FFDC6279]
       add       rsp,28
       pop       rsi
       pop       rbp
       ret
; Total bytes of code 175
```

> The caller of SetResult has to be generic non-inlineable method too for it to hit the inlining limitations.

The case I'm most interested in is the current use today, where the caller is the compiler-generated MoveNext method.  It's not a generic method, but it will be on a generic type _if_ the async method is generic.  I'm having trouble getting benchmark.net to show me the asm for a test of this case, so I'll need to build a checked coreclr with my changes....

Reply:
>  it will be on a generic type if the async method is generic

Yes, this is the case that I expect is going to hit the inlining limitations. It may be interesting to microbenchmark this case, in addition to looking at the disassembly, to see whether the missed inlining makes a difference.

Reply:
> 

Here's a benchmark of invoking an async method that does basically nothing, so this is purely about the overheads of the infrastructure and the worst case for this:
```C#
public class NonGeneric
{
    [Benchmark]
    public async Task<string> NonGenericWork()
    {
        for (int i = 0; i < 10000; i++) await GetValueAsync();
        return default;
    }

    private async Task<string> GetValueAsync() => default;
}

[GenericTypeArguments(typeof(string))]
[GenericTypeArguments(typeof(int))]
public class Generic<T>
{
    [Benchmark]
    public async Task<T> GenericWork()
    {
        for (int i = 0; i < 10000; i++) await GetValueAsync();
        return default;
    }

    private async Task<T> GetValueAsync() => default;
}
```

I get these numbers:

|            Type |         Method |        Job |           Toolchain |     Mean |   Error |  StdDev | Ratio | RatioSD |
|---------------- |--------------- |----------- |-------------------- |---------:|--------:|--------:|------:|--------:|
|  Generic<Int32> |    GenericWork | Job-HGAKDB | \master\corerun.exe | 142.7 us | 2.35 us | 3.36 us |  1.03 |    0.03 |
|  Generic<Int32> |    GenericWork | Job-KAESXO |     \pr\corerun.exe | 139.3 us | 0.67 us | 0.56 us |  1.00 |    0.00 |
|                 |                |            |                     |          |         |         |       |         |
| Generic<String> |    GenericWork | Job-HGAKDB | \master\corerun.exe | 227.0 us | 0.95 us | 0.79 us |  0.87 |    0.00 |
| Generic<String> |    GenericWork | Job-KAESXO |     \pr\corerun.exe | 260.6 us | 0.95 us | 0.79 us |  1.00 |    0.00 |
|                 |                |            |                     |          |         |         |       |         |
|      NonGeneric | NonGenericWork | Job-HGAKDB | \master\corerun.exe | 135.2 us | 1.15 us | 1.08 us |  1.03 |    0.01 |
|      NonGeneric | NonGenericWork | Job-KAESXO |     \pr\corerun.exe | 131.4 us | 0.42 us | 0.33 us |  1.00 |    0.00 |

So, the non-generic case and the value type case stay basically the same, maybe a smidgen better.  The generic case regresses by 15%.  But that is a super extreme.  While it's unfortunate the lack of inlining contributes such overhead here, I'm tempted to say it's acceptable. (Though it'd be wonderful of course if JIT improvements could remove it in the future.)

Reply:
> While it's unfortunate the lack of inlining contributes such overhead here, I'm tempted to say it's acceptable

+1